File: app/src/main/java/workshop/akbolatss/tagsnews/screen/reminders/ReminderService.java
package workshop.akbolatss.tagsnews.screen.reminders;

import android.app.AlarmManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;

import java.util.Calendar;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;
import me.toptas.rssconverter.RssFeed;
import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.api.NewsApiService;
import workshop.akbolatss.tagsnews.application.App;
import workshop.akbolatss.tagsnews.repositories.DBRssSourceRepository;
import workshop.akbolatss.tagsnews.repositories.source.RssSource;
import workshop.akbolatss.tagsnews.screen.splash.SplashActivity;

import static workshop.akbolatss.tagsnews.util.Constants.INTENT_HOUR;
import static workshop.akbolatss.tagsnews.util.Constants.INTENT_MINUTE;
import static workshop.akbolatss.tagsnews.util.Constants.INTENT_REQUEST_CODE;

public class ReminderService extends Service {

    public static final String TAG = "TAG";
    public static final int NOTIFICATION_ID = 99999;
    public static final String NOTIFICATION_CHANNEL_ID = "workshop.akbolatss.tagsnews.channel";
    public static final String NOTIFICATION_CHANNEL_NAME = "Tags News";
    public static final String NOTIFICATION_CHANNEL_DESCRIPTION = "Description of Tags News";

    @Inject
    protected NewsApiService mNewsApiService;

    @Inject
    protected DBRssSourceRepository mRepository;

    @Inject
    protected Context mContext;

    private CompositeDisposable mCompositeDisposable;

    private NotificationCompat.Builder builder;
    private NotificationCompat.BigTextStyle bigTextStyle;
    private NotificationCompat.InboxStyle inboxStyle;
    private String bigText;

    //Buffer Data
    private int mRepeateRequestCode;
    private int mRepeatHour;
    private int mRepeatMinute;

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        ((App) getApplicationContext()).getAppComponent().inject(this);

        mCompositeDisposable = new CompositeDisposable();

        builder = new NotificationCompat.Builder(mContext, getResources().getString(R.string.app_name));
        bigTextStyle = new NotificationCompat.BigTextStyle();
        inboxStyle = new NotificationCompat.InboxStyle();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        bigText = "";
        mRepeateRequestCode = intent.getIntExtra(INTENT_REQUEST_CODE, -1);
        mRepeatHour = intent.getIntExtra(INTENT_HOUR, -1);
        mRepeatMinute = intent.getIntExtra(INTENT_MINUTE, -1);

        mRepository.getOnlyActive()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new DisposableSingleObserver<List<RssSource>>() {
                    @Override
                    public void onSuccess(List<RssSource> rssSources) {
                        for (int i = 0; i < rssSources.size(); i++) {
                            final RssSource rssSource = rssSources.get(i);
                            mCompositeDisposable.add(mNewsApiService.getRss(rssSource.getLink())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribeOn(Schedulers.io())
                                    .subscribeWith(new DisposableSingleObserver<RssFeed>() {
                                        @Override
                                        public void onSuccess(RssFeed rssFeed) {
                                            String text = rssSource.getTitle() + " - " + rssFeed.getItems().get(0).getTitle();
                                            inboxStyle.addLine(text);

                                            builder.setSmallIcon(R.drawable.ic_icon)
                                                    .setContentTitle(getString(R.string.notification_title))
                                                    .setContentText(getString(R.string.notification_text))
                                                    .setAutoCancel(true)
                                                    .setWhen(System.currentTimeMillis())
                                                    .setStyle(inboxStyle)
                                                    .setChannelId(NOTIFICATION_CHANNEL_ID)
                                                    .setContentIntent(PendingIntent.getActivity(mContext, 0,
                                                            new Intent(mContext, SplashActivity.class), 0))
                                                    .setDefaults(Notification.DEFAULT_LIGHTS | Notification.DEFAULT_SOUND);

                                            NotificationManager notificationManager = (NotificationManager)
                                                    mContext.getSystemService(Context.NOTIFICATION_SERVICE);

                                            if (Build.VERSION.SDK_INT >= 26) {
                                                if (notificationManager.getNotificationChannel(NOTIFICATION_CHANNEL_ID) == null) {
                                                    int importance = NotificationManager.IMPORTANCE_LOW;

                                                    NotificationChannel mChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, importance);

                                                    // Configure the notification channel.
                                                    mChannel.setDescription(NOTIFICATION_CHANNEL_DESCRIPTION);

                                                    mChannel.enableLights(true);
                                                    // Sets the notification light color for notifications posted to this
                                                    // channel, if the device supports this feature.
                                                    mChannel.setLightColor(Color.GREEN);

                                                    mChannel.enableVibration(true);
                                                    mChannel.setVibrationPattern(new long[]{100, 200, 300, 400, 500, 400, 300, 200, 400});

                                                    notificationManager.createNotificationChannel(mChannel);
                                                }
                                            }

                                            notificationManager.notify(NOTIFICATION_ID, builder.build());
                                        }

                                        @Override
                                        public void onError(Throwable e) {
//                                            Log.d(TAG, "onError 2 " + e.getLocalizedMessage());
                                            onRepeatNotification();
                                        }
                                    }));
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
//                        Log.d(TAG, "onError 1 " + e.getLocalizedMessage());
                        onRepeatNotification();
                    }
                });

        return START_NOT_STICKY;
    }

    private void onRepeatNotification() {
        AlarmManager amc = (AlarmManager) mContext.getSystemService(ALARM_SERVICE);
        Intent myIntent = new Intent(mContext, ReminderReceiver.class);

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());

        int buffMinute = mRepeatMinute + 30;
        if (buffMinute >= 60) {
            mRepeatHour++;
            mRepeatMinute = buffMinute - 60;
            if (mRepeatHour >= 24) {
                mRepeatHour = 0;
            }
        } else {
            mRepeatMinute = buffMinute;
        }

        calendar.set(Calendar.HOUR_OF_DAY, mRepeatHour);
        calendar.set(Calendar.MINUTE, mRepeatMinute);

        myIntent.putExtra(INTENT_REQUEST_CODE, mRepeateRequestCode);
        myIntent.putExtra(INTENT_HOUR, mRepeatHour);
        myIntent.putExtra(INTENT_MINUTE, mRepeatMinute);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(mContext, mRepeateRequestCode, myIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        amc.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), pendingIntent);

        mContext.stopService(new Intent(mContext, ReminderService.class));
    }

    @Override
    public void onDestroy() {
        mCompositeDisposable.clear();
        super.onDestroy();
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/reminders/ReminderService.kt
package workshop.akbolatss.tagsnews.screen.reminders

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.job.JobParameters
import android.app.job.JobService
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.media.RingtoneManager
import android.os.Build
import android.support.v4.app.NotificationCompat

import javax.inject.Inject

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import me.toptas.rssconverter.RssFeed
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.api.NewsApiService
import workshop.akbolatss.tagsnews.application.App
import workshop.akbolatss.tagsnews.di.component.DaggerRemindersComponent
import workshop.akbolatss.tagsnews.di.module.RemindersModule
import workshop.akbolatss.tagsnews.model.DBRssSourceRepository
import workshop.akbolatss.tagsnews.model.dao.RssSource
import workshop.akbolatss.tagsnews.screen.splash.SplashActivity

import workshop.akbolatss.tagsnews.util.Constants.NOTIFICATION_CHANNEL_DESCRIPTION
import workshop.akbolatss.tagsnews.util.Constants.NOTIFICATION_CHANNEL_ID
import workshop.akbolatss.tagsnews.util.Constants.NOTIFICATION_CHANNEL_NAME
import workshop.akbolatss.tagsnews.util.Constants.NOTIFICATION_ID
import workshop.akbolatss.tagsnews.util.Logger
import java.util.*

class ReminderService : JobService() {

    @Inject
    lateinit var mNewsApiService: NewsApiService

    @Inject
    lateinit var mRepository: DBRssSourceRepository

    @Inject
    lateinit var mContext: Context

    private var mCompositeDisposable: CompositeDisposable? = null


    override fun onCreate() {
        super.onCreate()
        DaggerRemindersComponent.builder()
                .appComponent((application as App).appComponent)
                .remindersModule(RemindersModule())
                .build()
                .inject(this)

        mCompositeDisposable = CompositeDisposable()


    }

    override fun onStopJob(params: JobParameters?): Boolean {
        Logger.i("TAG", " Job Scheduler stopped")
        return false
    }

    override fun onStartJob(params: JobParameters?): Boolean {
//        mRepeatRequestCode = intent.getIntExtra(INTENT_REQUEST_CODE, -1)
//        mRepeatHour = intent.getIntExtra(INTENT_HOUR, -1)
//        mRepeatMinute = intent.getIntExtra(INTENT_MINUTE, -1)
        Logger.i("TAG", " Job Scheduler started")

        var counter = 0
        mRepository.onlyActive
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ rssSources ->
                    counter = rssSources.size
                    val inboxStyle = NotificationCompat.InboxStyle()

                    for (i in rssSources.indices) {
                        val rssSource = rssSources[i]
                        mCompositeDisposable!!.add(mNewsApiService.getRss(rssSource.link)
                                .observeOn(AndroidSchedulers.mainThread())
                                .subscribeOn(Schedulers.io())
                                .subscribe({ rssFeed: RssFeed ->
                                    counter++
                                    if (rssFeed.items.isNotEmpty()) {
                                        val text = rssSource.title + " - " + rssFeed.items[0].title
                                        inboxStyle.addLine(text)
                                        pushNotification(rssSource, rssFeed, inboxStyle)
                                    }
                                    if (counter >= rssSources.size)
                                        onStopJob(params)
                                }, {
                                    onStopJob(params)
                                }))
                    }
                }, {
                })
        return true
    }


    private fun pushNotification(rssSource: RssSource, rssFeed: RssFeed, inboxStyle: NotificationCompat.InboxStyle) {



        val icon = BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher)

        val intent = Intent(mContext, SplashActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        val pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT)
        val notificationBuilder = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
                .setContentTitle(getString(R.string.notification_title))
                .setContentText(getString(R.string.notification_text))
                .setAutoCancel(true)
                .setChannelId(NOTIFICATION_CHANNEL_ID)
                .setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION))
                .setLargeIcon(icon)
                .setContentIntent(pendingIntent)
                .setStyle(inboxStyle)
                .setColor(Color.GREEN)
                .setLights(Color.GREEN, 1000, 2000)
                .setDefaults(Notification.DEFAULT_LIGHTS or Notification.DEFAULT_SOUND)
                .setSmallIcon(R.mipmap.ic_launcher)

        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT)

            channel.description = NOTIFICATION_CHANNEL_DESCRIPTION
            channel.setShowBadge(true)
            channel.canShowBadge()
            channel.enableLights(true)
            channel.lightColor = Color.GREEN
            channel.enableVibration(true)
            channel.vibrationPattern = longArrayOf(100, 200, 300, 400, 500)
            notificationManager.createNotificationChannel(channel)
        }

//        val random = Random()
//        val m = random.nextInt(9999 - 1000) + 1000
        notificationManager.notify(NOTIFICATION_ID, notificationBuilder.build())
    }

    override fun onDestroy() {
        mCompositeDisposable!!.clear()
        super.onDestroy()
    }
}
